File: wakakusa/demo/styled.rb
# A look kept in one place: a Hash of properties, merged and handed to
# an element with `**`. `theme:` swaps the palette its subtree resolves
# colors in, so one keyword flips the whole panel.
require "wakakusa"

CHIP = { size: 18.0, color: "accent" }.freeze
KEY = { background: "#313244", hover_background: "#45475a" }.freeze
HOT = { background: "#fab387" }.freeze
KEY_HOT = KEY.merge(HOT).freeze

class Styled
  def initialize
    @mode = "dark"
    @n = 0
  end

  def flip
    @mode = @mode == "dark" ? "light" : "dark"
  end

  def view
    column(
      text("n=#{@n}", **CHIP),
      row(
        button("+1", **KEY) { @n += 1 },
        button("flip", **KEY_HOT) { flip },
        spacing: 6.0
      ),
      spacing: 8.0, padding: 12.0, background: "panel", theme: @mode
    )
  end
end

run(Styled.new, title: "styled")

File: gomamochi/demo/styled.go
// A look kept in one place: a function that sets the same properties
// on every button handed to it. `Theme` swaps the palette its subtree
// resolves colors in, so one keyword flips the whole panel.
package main

import (
	"fmt"

	. "github.com/i2y/yokan/gomamochi"
)

func key(b *ButtonEl) *ButtonEl { return b.Background("#313244").HoverBackground("#45475a") }

func keyHot(b *ButtonEl) *ButtonEl { return key(b).Background("#fab387") }

type Styled struct {
	mode string
	n    int
}

func (s *Styled) flip() {
	if s.mode == "dark" {
		s.mode = "light"
	} else {
		s.mode = "dark"
	}
}

func (s *Styled) View() Element {
	return Column(
		Text(fmt.Sprintf("n=%d", s.n)).Size(18).Color("accent"),
		Row(
			key(Button("+1")).OnClick(func() { s.n += 1 }),
			keyHot(Button("flip")).OnClick(func() { s.flip() }),
		).Spacing(6),
	).Spacing(8).Padding(12).Background("panel").Theme(s.mode)
}

func main() {
	Run(&Styled{mode: "dark"}, Title("styled"))
}
